### PR TITLE
Automatic versioning with versioneer

### LINKED ISSUE #4727: Use something like `versioneer`
Instead of manually updating a version string in code, why don't we use something like [`versioneer`](https://github.com/python-versioneer/python-versioneer) that allows us to utilize `git` tags?

### PR DESCRIPTION
**Pull request by ghost**
This adds automatic versioning using the versioneer package.

**Checklist**
+ [x] Explain important implementation details 👆
+ [x] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [x] Link relevant issues (preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html))
+ [x] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇

## Major / Breaking Changes
- ...

## Bugfixes / New features
- Closes #4727

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -11,3 +11,10 @@ convention = numpy
 [isort]
 lines_between_types = 1
 profile = black
+
+[versioneer]
+VCS = git
+style = pep440
+versionfile_source = pymc/_version.py
+versionfile_build = pymc/_version.py
+tag_prefix = rel-
```
Comment (brandonwillard) [inline setup.cfg]: You might want to use `tag_prefix = v` to match the current format of PyMC's release version tags.

**Comment by twiecki** [comment]
image
Glad to see new and promising contributors emerge! ;)

**Comment by ghost** [comment]
@twiecki yeah, I used my work account by mistake!

**Comment by twiecki** [comment]
Where do these mypy errors come from? CC @michaelosthege

**Comment by michaelosthege** [comment]
> Where do these mypy errors come from? CC @michaelosthege

Please just read the output of the mypy step. It says:

```
57/73 files pass as expected.
!!!!!!!!!
1 files unexpectedly passed the type checks:
pymc/_version.py
This is good news! Go to scripts/run_mypy.py and add them to the list.
!!!!!!!!!
```

But there are a bunch of black/pylint issues in that last pre-commit run.

**Comment by fonnesbeck** [comment]
@OriolAbril any idea why the RTD would be failing here? Does not seem to be related to versioning.

**Comment by OriolAbril** [comment]
I don't know, I'd have to build the docs locally to access all the logs. It might be related to the latest release of the theme. Try setting `pydata-sphinx-theme<0.10.0` in the conda envs and if it works, open an issue so we remember to look into the actual error.

**Comment by ghost** [comment]
Yeah, the local build of docs is failing for me. Its looking for a lot of API stuff that does not exist anymore, like `pandas_to_array`, `AR1`, etc.

```
WARNING: [autosummary] failed to import pymc.AR1.
Possible hints:
* AttributeError: module 'pymc' has no attribute 'AR1'
* ModuleNotFoundError: No module named 'pymc.AR1'
* ImportError: 
WARNING: [autosummary] failed to import pymc.EulerMaruyama.
Possible hints:
* ModuleNotFoundError: No module named 'pymc.EulerMaruyama'
* ImportError: 
* AttributeError: module 'pymc' has no attribute 'EulerMaruyama'
WARNING: [autosummary] failed to import pymc.KullbackLiebler.
Possible hints:
* AttributeError: module 'pymc' has no attribute 'KullbackLiebler'
* ImportError: 
* ModuleNotFoundError: No module named 'pymc.KullbackLiebler'
WARNING: [autosummary] failed to import pymc.NFVI.
Possible hints:
* ImportError: 
* ModuleNotFoundError: No module named 'pymc.NFVI'
* AttributeError: module 'pymc' has no attribute 'NFVI'
WARNING: [autosummary] failed to import pymc.NormalizingFlow.
Possible hints:
* ModuleNotFoundError: No module named 'pymc.NormalizingFlow'
* AttributeError: module 'pymc' has no attribute 'NormalizingFlow'
* ImportError: 
WARNING: [autosummary] failed to import pymc.SimulatorRV.
Possible hints:
* ImportError: 
* ModuleNotFoundError: No module named 'pymc.SimulatorRV'
* AttributeError: module 'pymc' has no attribute 'SimulatorRV'
WARNING: [autosummary] failed to import pymc.compile_pymc.
Possible hints:
* ImportError: 
* AttributeError: module 'pymc' has no attribute 'compile_pymc'
* ModuleNotFoundError: No module named 'pymc.compile_pymc'
WARNING: [autosummary] failed to import pymc.math.abs_.
Possible hints:
* AttributeError: module 'pymc.math' has no attribute 'abs_'
* ImportError: 
* ModuleNotFoundError: No module named 'pymc.math.abs_'; 'pymc.math' is not a package
WARNING: [autosummary] failed to import pymc.pandas_to_array.
Possible hints:
* AttributeError: module 'pymc' has no attribute 'pandas_to_array'
* ImportError: 
* ModuleNotFoundError: No module named 'pymc.pandas_to_array'
```

**Comment by twiecki** [comment]
`abs_` was renamed to `abs`, is it still mentioned in the docs?

Not sure where the other ones are coming from. In any case, they are not related to this PR, so maybe we should fix that separately and merge this in.

**Comment by OriolAbril** [comment]
These are all only warnings which have always been around since the first v4 doc build, they fill the screen with red but don't break the website generation process. We also have an issue to track this #5282 because getting rid of those is easy but the symmetric case also happens, there are still many functions that are new instead of removed and do not appear on the docs, and or that there is no warning.

**Comment by twiecki** [comment]
So can we merge this safely?

**Comment by OriolAbril** [comment]
The build error should have been fixed in #6120 and builds for latest are working: https://readthedocs.org/projects/pymc/builds/ so it is now safe to merge this one even if rtd didn't work. We did have more than a week where the latest tag in the docs didn't match github as the rtd build failed and it could not be updated.

**Comment by ricardoV94** [comment]
@twiecki seems like you didn't squash this?

**Comment by twiecki** [comment]
Yes :(.

**Comment by ricardoV94** [comment]
> Yes :(.

Oh well, time to start from scratch. PyMC5?

**Comment by OriolAbril** [comment]
Should we bring back the number as part of the package name too? 😜

**Comment by michaelosthege** [comment]
I've added the `major` label, because this changes how the `pm.__version__` looks like.
Whoever makes the release please explain/link how to deal with version number comparisons for installs from GitHub =)

**Comment by maresb** [comment]
Are we sure that `versioneer` needs to be added to the install requirements? This doesn't look right to me.

**Comment by maresb** [comment]
Regarding my last comment, I started an experiment at #6135

**Pull request merged**